src/main.c

```diff
@@ -54,6 +54,11 @@ char *bshell_read_line() {
     // Read a character
     c = getchar();
 
+    // Ctrl+D makes the shell loop infinitely.
+    // A quick printf reveals that ctrl+d makes c equal to -1.
+    if (c == -1) {
```

Comment:
It's not guaranteed to return -1, it returns EOF (defined in stdio.h), you should check for that instead (and remove the check a few lines down).